subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c

```diff
@@ -0,0 +1,542 @@
+		confirmed = false;
+		pending = false;
+		permanent = false;
+		img_num = 0;
+		slot_num = -1;
+		hash.len = 0;
+		version.len = 0;
+
+		rc = zcbor_map_decode_bulk(zsd, list_res_decode, ARRAY_SIZE(list_res_decode),
+					   &decoded);
+		if (rc) {
+			if (slot) {
+				/* No more map */
+				break;
+			}
+			LOG_ERR("Corrupted Image data %d", rc);
+			response.status = MGMT_ERR_ECORRUPT;
+			rc = MGMT_ERR_EINVAL;
+			goto out;
+		}
+		/* Check that mandatory parametrs have decoded */
+		if (hash.len != IMG_MGMT_HASH_LEN || !version.len || slot_num == -1) {
+			LOG_ERR("Missing mandatory parametrs");
+			response.status = MGMT_ERR_ECORRUPT;
+			rc = MGMT_ERR_EINVAL;
+			goto out;
+		}
+		/* Store parsed values */
+		image_info[image_count].img_num = img_num;
+		image_info[image_count].slot_num = slot_num;
+		memcpy(image_info[image_count].hash, hash.value, IMG_MGMT_HASH_LEN);
+		if (version.len >= IMG_MGMT_VER_MAX_STR_LEN) {
+			version.len = IMG_MGMT_VER_MAX_STR_LEN;
+		}
+		memcpy(image_info[image_count].version, version.value, version.len);
+		image_info[image_count].version[version.len] = '\0';
+		image_info[image_count].bootable = bootable;
+		image_info[image_count].pending = pending;
+		image_info[image_count].confirmed = confirmed;
+		image_info[image_count].active = active;
+		image_info[image_count].permanent = permanent;
+		/* Update valid image count */
+		image_count++;
+	}
+
+	ok = zcbor_list_end_decode(zsd);
+	if (!ok) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
+	} else {
+		response.image_state.image_list_length = image_count;
+		response.image_state.image_list = image_info;
+		response.status = MGMT_ERR_EOK;
+		rc = 0;
+	}
+
+out:
+	if (response.status != MGMT_ERR_EOK) {
+		image_count = 0;
+	}
+	client_gr_image_fn(&response);
+	k_sem_give(user_data);
+	return rc;
+}
+
+static int image_upload_res_fn(struct net_buf *nb, void *user_data)
+{
+	struct mcumgr_client_image_gr_rsp response;
+	zcbor_state_t zsd[CBOR_ENCODER_STATE_NUM + 2];
+	size_t decoded;
+	int rc;
+
+	response.cmd = IMG_MGMT_ID_UPLOAD;
+	response.status = -1;
+	response.image_upload_offset = SIZE_MAX;
+
+	struct zcbor_map_decode_key_val upload_res_decode[] = {
+		ZCBOR_MAP_DECODE_KEY_DECODER("off", zcbor_size_decode,
+					     &response.image_upload_offset),
+		ZCBOR_MAP_DECODE_KEY_DECODER("rc", zcbor_int32_decode, &response.status)};
+
+	if (!nb) {
+		response.status = MGMT_ERR_ETIMEOUT;
+		upload_state.upload_status = MGMT_ERR_ETIMEOUT;
+		client_gr_image_fn(&response);
+		k_sem_give(user_data);
+		return 0;
+	}
+
+	zcbor_new_decode_state(zsd, ARRAY_SIZE(zsd), nb->data, nb->len, 1);
+
+	rc = zcbor_map_decode_bulk(zsd, upload_res_decode, ARRAY_SIZE(upload_res_decode), &decoded);
+	if (rc || response.status == -1 || response.image_upload_offset == SIZE_MAX) {
+		response.status = MGMT_ERR_ECORRUPT;
+		upload_state.upload_status = MGMT_ERR_ECORRUPT;
+		client_gr_image_fn(&response);
+		k_sem_give(user_data);
+		return MGMT_ERR_EINVAL;
```

Comment:
Instead of doing this a lot, have a `goto` e.g. `goto end;` then have the `end:` part where it gives the semaphore and returns e.g. `rc`